catalyst_voices/packages/libs/catalyst_cardano_serialization/lib/src/builders/input_builder.dart

```diff
@@ -0,0 +1,296 @@
+  ///     outputs and minimum ADA requirements.
+  ///
+  /// - Returns:
+  ///   A tuple containing:
+  ///   - [List<TransactionOutput>]: The list of change outputs.
+  ///   - [Coin]: The total transaction fee, including the fee for the change
+  ///     outputs.
+  ///   Returns `null` if the accumulated balance is insufficient to cover the
+  ///   required amount and fees.
+  ///
+  /// - Notes:
+  ///   - This method ensures the transaction remains valid by confirming that
+  ///     the accumulated balance exceeds or equals the required amount plus
+  ///     fees.
+  (List<TransactionOutput>, Coin)? _getChangeAndFee(
+    AssetId assetId,
+    TransactionBuilder builder,
+    Set<TransactionUnspentOutput> selectedInputs,
+    Balance selectedTotal,
+    Balance targetTotal,
+  ) {
+    final minFee =
+        builder.copyWith(inputs: selectedInputs).minFee(useWitnesses: true);
+    final minimumRequired = targetTotal + Balance(coin: minFee);
+
+    if (selectedTotal.lessThan(minimumRequired)) return null;
+
+    // Calculate the change to be returned, as the selected total is guaranteed
+    // to be larger.
+    final changeTotal = selectedTotal - minimumRequired;
+
+    final (changeOutputs, changeFee) = _buildChangeOutputs(
+      builder,
+      changeTotal,
+      minFee,
+    );
+
+    final requiredFee = minFee + changeFee;
+    if (selectedTotal.lessThan(targetTotal + Balance(coin: requiredFee))) {
+      return null;
+    }
+
+    return (changeOutputs, requiredFee);
+  }
+
+  /// Constructs the change outputs and calculates their associated fees.
+  ///
+  /// This method generates the necessary change outputs based on the remaining
+  /// balance after covering the required amount and transaction fees. It also
+  /// calculates the additional fee incurred by the change outputs.
+  ///
+  /// - Parameters:
+  ///   - [builder]: The transaction builder used to create the transaction.
+  ///   - [remainingBalance]: The balance remaining after deducting the required
+  ///     amount and initial fees.
+  ///   - [minFee]: The minimum fee for the transaction before considering
+  ///     change outputs.
+  ///
+  /// - Returns:
+  ///   A tuple containing:
+  ///   - [List<TransactionOutput>]: The list of change outputs.
+  ///   - [Coin]: The additional fee incurred by the change outputs.
+  ///
+  /// - Notes:
+  ///   - If the remaining balance is zero, no change outputs are created, and
+  ///     the fee remains unchanged.
+  ///   - This method handles multi-asset balances and ensures they are included
+  ///     in the change outputs, if applicable.
+  (List<TransactionOutput>, Coin) _buildChangeOutputs(
+    TransactionBuilder builder,
+    Balance remainingBalance,
+    Coin minFee,
+  ) {
+    if (remainingBalance.isZero) return ([], const Coin(0));
+    if (builder.changeAddress == null) {
+      throw Exception('Change address required for non-zero balance.');
```

Comment:
I'd change it to an `Error` because it's programmer's mistake to not pass the change address.

```suggestion
      throw ArgumentError('Change address required for non-zero balance.');
```

Reply:
done